### PR TITLE
Add support for generating message type registry

### PR DESCRIPTION
**Pull request by aikoven**
This PR adds support for generating type registry.

Type registry is a `Map` defined in `typeRegistry.ts` file, that maps fully-qualified message type name to an object containing static message methods (`encode`, `decode`, `fromPartial` etc.).

Also, each message interface gets an extra `$type` field containing fully-qualified message type name.

### CODE REVIEW

File: integration/type-registry/bar/bar.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+export const protobufPackage = 'foo.bar';
+
+export interface Bar {
+  $type: 'foo.bar.Bar';
+  foo: Foo | undefined;
+}
+
+const baseBar: object = { $type: 'foo.bar.Bar' };
+
+export const Bar = {
```
Comment (aikoven) [inline integration/type-registry/bar/bar.ts:16]: I think this object should have a `$type` field as well

Comment (stephenh) [reply integration/type-registry/bar/bar.ts:16]: Sgtm

File: src/main.ts [hunk 2]
```diff
@@ -132,6 +142,14 @@ export function generateFile(ctx: Context, fileDesc: FileDescriptorProto): [stri
             ${joinCode(staticMethods, { on: ',\n\n' })}
           };
         `);
+
+        if (options.outputTypeRegistry) {
+          const messageTypeRegistry = imp('messageTypeRegistry@./typeRegistry');
+
+          chunks.push(code`
+            ${messageTypeRegistry}.set('${fullTypeName}', ${def(fullName)});
```
[the remark this answers is not in the document]
Comment (aikoven) [reply src/main.ts:150]: Do I understand correctly, that `def` always keeps the original name? And only `imp` can do a rename?

File: integration/type-registry/bar/bar.ts (hunk 1 above)
Comment (aikoven) [reply integration/type-registry/bar/bar.ts:16]: Added the `$type` field to message statics.

This should later help with `google.protobuf.Any` helpers:

```ts
const anyMessage: Any = ...;

if (Any.is(anyMessage, Foo)) {
  const foo: Foo = Any.unpack(anyMessage, Foo);
}
```

https://github.com/protocolbuffers/protobuf/blob/v3.15.6/src/google/protobuf/any.proto#L67-L75

File: src/main.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (aikoven) [reply src/main.ts]: Removed the `def`.

File: src/generate-type-registry.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+  return joinCode(chunks, { on: '\n\n' });
+}
+
+function generateMessageType(ctx: Context): Code {
+  const chunks: Code[] = [];
+
+  chunks.push(code`export interface MessageType<Message> {`);
+
+  if (ctx.options.outputEncodeMethods) {
```
[the remark this answers is not in the document]
Comment (aikoven) [reply src/generate-type-registry.ts:26]: Thanks. I also ran all the existing integration tests against the new option, everything generates well with no typescript errors.

File: src/main.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (stephenh) [reply src/main.ts]: Yeah, you're right that `imp`s will be renamed, and `def`s won't , but `def` is short for `definition` and so is for declaring the symbols that `imp` should be renamed around.

I.e. if I'm `imp`orting a symbol `Foo`, but I also have a `class Foo` in my file's output, if I wrap that definition of foo as `class def(Foo)` it will tell the auto-import logic that `Foo` is taken and to rename the imported to `Foo1` or what not.

Ah yeah, so for here we're referring to (say) `FooMessage` but we can trust/know that the `export const FooMessage = { ... }` already marked/wrapped its definition/declaration of `FooMessage` in a `def`.

Basically we only need one `def` per symbol per file, ideally at that symbols declaration point.

File: src/plugin.ts [hunk 4]
```diff
@@ -29,6 +30,18 @@ async function main() {
       return { name: path, content: prefixDisableLinter(spec) };
     })
   );
+
+  if (options.outputTypeRegistry) {
+    const utils = makeUtils(options);
+    const ctx: Context = { options, typeMap, utils };
```
[the remark this answers is not in the document]
Comment (stephenh) [reply src/plugin.ts:36]: Yep, that comment is what I was half-remembering, so what you have lgtm; glad that it reminded future us of that rationale. :-)

File: integration/type-registry/bar/bar.ts (hunk 1 above)
Comment (stephenh) [reply integration/type-registry/bar/bar.ts:16]: Nice!

**Comment by aikoven** [comment]
Thank you for your review! You are a great maintainer, and this is a really nice library.

Just in case, I can't merge this PR myself after your approval, there's no merge button.

**Comment by stephenh** [comment]
> Just in case, I can't merge this PR myself after your approval, there's no merge button.

I was wondering about that! :-)

Cool, I'll hit merge and get a release out in the next day or so; I'm batching together a few PRs. Also want to get an "npm release automatically on PR merge" going.
